source/server/hot_restarting_child.cc

```diff
@@ -154,11 +155,33 @@ void HotRestartingChild::registerUdpForwardingListener(
   udp_forwarding_context_.registerListener(address, listener_config);
 }
 
+void HotRestartingChild::registerParentDrainedCallback(
+    const Network::Address::InstanceConstSharedPtr& address, absl::AnyInvocable<void()> callback) {
+  if (parent_drained_) {
+    callback();
+  } else {
+    auto result = on_drained_actions_.try_emplace(address->asString(), std::move(callback));
+    RELEASE_ASSERT(result.second,
+                   fmt::format("two udp listeners with the same address ({}) is unexpected",
+                               address->asString()));
+  }
+}
+
+void HotRestartingChild::allDrainsImplicitlyComplete() {
+  for (auto& drain_action : on_drained_actions_) {
+    // call the callback.
+    std::move(drain_action.second)();
+  }
+  on_drained_actions_.clear();
+  parent_drained_ = true;
+}
+
 absl::optional<HotRestart::AdminShutdownResponse>
 HotRestartingChild::sendParentAdminShutdownRequest() {
-  if (restart_epoch_ == 0 || parent_terminated_) {
+  if (parent_terminated_) {
     return absl::nullopt;
   }
+  allDrainsImplicitlyComplete();
```

Comment:
I don't think drains have completed at this time. This is called relatively early in child process startup, before draining of the parent has happened. At least I think so.

Reply:
Good catch, I misunderstood what this was. Changed to only being drain-complete on terminate.